internalshared/configutil/http_response_headers.go

```diff
@@ -0,0 +1,186 @@
+        if _, ok := c[hn]; ok {
+            continue
+        }
+        hv := ParseDefaultHeaders(hn)
+        if hv != "" {
+            defaults[hn] = hv
+        }
+    }
+
+    return defaults
+}
+
+func ParseCustomResponseHeaders(r interface{}) (map[string]map[string]string, error) {
+    if !isValidListDict(r) {
+        return nil, fmt.Errorf("response headers were not configured correctly. please make sure they're in a map")
+    }
+
+    customResponseHeader := r.([]map[string]interface{})
+    h := make(map[string]map[string]string)
+
+    for _, crh := range customResponseHeader {
+        for sc, rh := range crh {
+            if !isValidListDict(rh){
+                return nil, fmt.Errorf("invalid response header type")
+            }
+
+            if !IsValidStatusCode(sc) {
+                return nil, fmt.Errorf("invalid status code found in the config file: %v", sc)
+            }
+
+            hvl := rh.([]map[string]interface{})
+            if len(hvl) != 1 {
+                return nil, fmt.Errorf("invalid number of response headers exist")
+            }
+            hvm := hvl[0]
+            hv, err := parseHeaders(hvm)
+            if err != nil {
+                return nil, err
+            }
+
+            h[sc] = hv
+        }
+    }
+
+    // setting default custom headers
+    de := h["default"]
+    h["default"] = setDefaultResponseHeaders(de)
+
+	return h, nil
+}
+
+func isValidListDict(in interface{}) bool {
+    if _, ok := in.([]map[string]interface{}); ok {
+        return true
+    }
+    return false
+}
+
+func isValidList(in interface{}) bool {
+    if _, ok := in.([]interface{}); ok {
+        return true
+    }
+    return false
+}
+
+func IsValidStatusCodeCollection(sc string) bool {
+    for _, v := range ValidCustomStatusCodeCollection {
+        if sc == v {
+            return true
+        }
+    }
+
+    return false
+}
+
+// IsValidStatusCode checking for status codes outside the boundary
+func IsValidStatusCode(sc string) bool {
+    if IsValidStatusCodeCollection(sc) {
+        return true
+    }
+
+    i, err := strconv.Atoi(sc)
+    if err != nil {
+       return false
+    }
+
+    if i >= 600 || i < 100 {
+        return false
+    }
+
+    return true
+}
+
+func parseHeaders(in map[string]interface{}) (map[string]string, error) {
+    hvMap := make(map[string]string)
+    for k, v := range in {
+        // parsing header name
+        hn := textproto.CanonicalMIMEHeaderKey(k)
+        // parsing header values
+        s, err := parseHeaderValues(v)
+        if err != nil {
+            return nil, err
+        }
+        hvMap[hn] = s
+    }
+    return hvMap, nil
+}
+
+func parseHeaderValues(h interface{}) (string, error) {
+    var sl []string
+    if !isValidList(h) {
+        return "", fmt.Errorf("headers must be given in a list of strings")
+    }
+    vli := h.([]interface{})
+    for _, vh := range vli {
+        sl = append(sl, vh.(string))
```

Comment:
I wonder if we should do any safety checks on the strings, e.g. to make sure they don't contain any invalid header chars like space or semicolon?

Reply:
we could do that, however, one thing to consider is that operators may want to just pass in one string with all headers specified or pass in a list of headers values which comprise couple of headers per item. Those could be have spaces and semicolons. We have examples of it in /ui and also I have specified default headers the same way. For example:
strictTransportSecurity = "max-age=31536000; includeSubDomains"
So, searching for semicolon and spaces may cause issues. what do you think? is there any other particular invalid chars that we need to care for?

Reply:
Ah, fair enough.  No, let's leave it be then.